### PR TITLE
External libraries preview column

### PR DESCRIPTION
**Pull request by daschuer**
this PR gives all external libraries a preview column.

### CODE REVIEW

**Comment by daschuer** [comment]
Just fixed: https://bugs.launchpad.net/mixxx/+bug/1197633

**Comment by ywwg** [comment]
can this method also be used to create a Played column?

**Comment by daschuer** [comment]
Fixed: https://bugs.launchpad.net/mixxx/+bug/1160525

**Comment by daschuer** [comment]
> can this method also be used to create a Played column?

No, because it needs a real database table.

**Comment by esbrandt** [comment]
Thanks, works nice with OSX and latest master.
Tested with iTunes and Traktor as external libraries.

Two minor things:
* ``Preview`` column is on last position in ``Browse`` mode header, while on 1st ( or 2nd) position everywhere else.
* ``Preview`` column icon looses active state, if you don`t hover with the mouse over the icon, even if the track is still previewing.

**Comment by daschuer** [comment]
Thank for for review!
@ywwg: Is there an issue when I change the header default order?

**Comment by daschuer** [comment]
@esbrandt: There was an issue that the playing track was not recognized because of the foreign track IDs. In the browse model, the previewing state was not tracked.
It is now both fixed. The preview column is in the leftmost position of the browse table now.

**Comment by esbrandt** [comment]
Looks good so far.
But now i get a bug, where if i start previewing a track in a external library, a warning window pops up that blocks all interactions with the GUI.

``Could not update metadata of track  "/Users/foo/Music/Mixxx/Recordings/2015-01-23_14h36m54s.wav``.

The pop-up can not be clicked away, and i have to terminate the Mixxx session.

log reads
```bash
Debug [Main]: BrowseTableModel::setData( QVariant(QString, "0") ) 
Debug [Main]: Successfully updated metadata of track  "/Users/foo/Music/Mixxx/Recordings/2015-01-20_16h44m57s.wav" 
Debug [Main]: BrowseTableModel::setData( QVariant(QString, "0") ) 
Debug [Main]: Could not update metadata of track  "/Users/foo/Music/Mixxx/Recordings/2015-01-23_14h36m54s.wav" 
```
The ``*.wav`` file in question is the 1st part (650055724 bytes =650.1 MB) of a split size recording .

**Comment by daschuer** [comment]
@esbrandt: Please test again.

**Comment by esbrandt** [comment]
The issue with the warning-popup is gone, thanks.

Found some more quirks in ``Browse`` mode:
Clicking Preview triggers a sorting,  e.g. preview the topmost track and that track moves to the bottom.
More troublesome is that if you preview multiple tracks after another in that directory, Mixxx GUI will stall (preview still audible) and you have to force exit. This happens mostly after the 5th but sometimes when previewing the last track of a directory.

**Comment by daschuer** [comment]
This seams to be difficult, sorry.
Should work now.

**Comment by esbrandt** [comment]
Almost done:)
GUI stalling is gone, but you can still trigger a sorting when previewing.
Just have the preview header active in the browse table, and every previewed track slides to the bottom. This is just with the ``Browse`` mode.

**Comment by daschuer** [comment]
The question here is bug of feature. When I tested it I thought it could be a feature to find the previewed track.
Sorting is now at the file name bay default. In your case, It might stick on the Preview column due to the stored header infos. 
But since this feature is not implemented elsewhere, I will remove it here as well.

**Comment by esbrandt** [comment]
Hmm, ok i can see that sorting might be useful when clicking on the Preview decks header works like with every other column header. But if, it should work the same way in all supported preview columns, not only in Browse mode. I think it is the right way to just disable sorting for now.

**Comment by daschuer** [comment]
sorting disabled.

File: src/playerinfo.cpp [hunk 1]
```diff
@@ -57,12 +57,15 @@ TrackPointer PlayerInfo::getTrackInfo(const QString& group) {
 }
 
 void PlayerInfo::setTrackInfo(const QString& group, const TrackPointer& track) {
-    QMutexLocker locker(&m_mutex);
-    TrackPointer pOld = m_loadedTrackMap.value(group);
+    TrackPointer pOld;
+    { // Scope
+        QMutexLocker locker(&m_mutex);
+        TrackPointer pOld = m_loadedTrackMap.value(group);
```
Comment (rryan) [inline src/playerinfo.cpp]: local pOld hides the parent score pOld

**Comment by daschuer** [comment]
Fixed!

File: src/library/trackmodel.h [hunk 2]
```diff
@@ -131,6 +131,11 @@ class TrackModel {
         m_eDefaultSortOrder = sortOrder;
     }
 
+    virtual bool getIsClolumnSortable(int column) {
```
Comment (rryan) [inline src/library/trackmodel.h]: Clolumn -> Column

Comment (rryan) [inline src/library/trackmodel.h]: Also getIs is slightly redundant -- mabye ```isColumnSortable``` or just ```columnSortable``` ?

**Comment by daschuer** [comment]
Fixed

**Comment by esbrandt** [comment]
After testing the PR for 2 weeks with no major issues on OSX and iTunes/Traktor libraries, there is only a tiny cosmetic thing left.
Rarely, if you do (secondary) sorting by clicking in the headers, you can end up with 2 or more active icons in the preview column. If the previewed track reached the end, the phantom icons stay active, you can disable them by clicking. Pattern is similar to: Preview random track > Sort header (e.g. Artist) > Preview next track
mixxx

Other than that, good to go for me.

File: src/library/browse/browsethread.cpp [hunk 3]
```diff
@@ -232,12 +236,12 @@ void BrowseThread::populateModel() {
         if (row % 10 == 0) {
             // this is a blocking operation
             emit(rowsAppended(rows, thisModelObserver));
-            //qDebug() << "Append " << rows.count() << " from " << filepath;
```
Comment (rryan) [inline src/library/browse/browsethread.cpp]: could you turn off the qdebugs?

File: src/library/browse/browsetablemodel.cpp [hunk 4]
```diff
@@ -389,8 +401,39 @@ bool BrowseTableModel::setData(const QModelIndex &index, const QVariant &value,
     }
 }
 
+void BrowseTableModel::trackLoaded(QString group, TrackPointer pTrack) {
```
Comment (rryan) [inline src/library/browse/browsetablemodel.cpp]: nit: const reference

File: src/library/baseexternalplaylistmodel.cpp [hunk 5]
```diff
@@ -131,6 +131,35 @@ void BaseExternalPlaylistModel::setPlaylist(QString playlist_path) {
     setSearch("");
 }
 
+void BaseExternalPlaylistModel::trackLoaded(QString group, TrackPointer pTrack) {
```
Comment (rryan) [inline src/library/baseexternalplaylistmodel.cpp]: nit: some way to avoid the code duplication?

**Comment by rryan** [comment]
code overall LGTM but it looks like @esbrandt has hit a bug?

**Comment by daschuer** [comment]
@esbrandt: I have tried hard to reproduce the issue on my Ubuntu Trusty using the Rhythmbox library, but I failed. I have fixed a minor issue with refreshing in my last commit, but it makes no difference to me. 
Please check it on your system. 

Since the loaded state is a single integer, that holds the track ID, there can't go much wrong. 
I can only think of two cases, which we can sort out: 
* different tracks with the same ID -> than the issue cannot appear only after sorting. 
* A refresh issue -> The whole library is refreshed when the play state of the preview deck changed.
This can also not be the issue. 

Is this a general issue for Mixxx play-lists and library view as well? Is only iTunes or Traktor effected? 
Are only external playlists effected or also the external library view (root folder) 
Any other obvious points?

**Comment by esbrandt** [comment]
Thanks for looking.
I`ll test later today and let you know.

**Comment by daschuer** [comment]
Any Updates? Since this PR fixes more than it brakes. can we merge it and fix the remaining bug later once we know how?

**Comment by kain88-de** [comment]
@esbrandt do you still see the bug?

**Comment by rryan** [comment]
let's fix any issues as they come up

**Pull request merged**